test/lib/cql_test_env.cc

```diff
@@ -594,6 +598,62 @@ class single_node_cql_env : public cql_test_env {
+            auto stop_sst_dir_sem = defer([&sst_dir_semaphore] {
+                sst_dir_semaphore.stop().get();
+            });
+
+            replica::database_config dbcfg;
+            if (cfg_in.dbcfg) {
+                dbcfg = std::move(*cfg_in.dbcfg);
+            } else {
+                dbcfg.available_memory = memory::stats().total_memory();
+            }
+
+            dbcfg.compaction_scheduling_group = scheduling_groups.compaction_scheduling_group;
+            dbcfg.memory_compaction_scheduling_group = scheduling_groups.memory_compaction_scheduling_group;
+            dbcfg.streaming_scheduling_group = scheduling_groups.streaming_scheduling_group;
+            dbcfg.statement_scheduling_group = scheduling_groups.statement_scheduling_group;
+            dbcfg.memtable_scheduling_group = scheduling_groups.memtable_scheduling_group;
+            dbcfg.memtable_to_cache_scheduling_group = scheduling_groups.memtable_to_cache_scheduling_group;
+            dbcfg.gossip_scheduling_group = scheduling_groups.gossip_scheduling_group;
+            dbcfg.sstables_format = sstables::version_from_string(cfg->sstable_format());
+
+            auto get_tm_cfg = sharded_parameter([&] {
+                return tasks::task_manager::config {
+                    .task_ttl = cfg->task_ttl_seconds,
+                };
+            });
+            task_manager.start(std::move(get_tm_cfg), std::ref(abort_sources)).get();
+            auto stop_task_manager = defer([&task_manager] {
+                task_manager.stop().get();
+            });
+
+            // get_cm_cfg is called on each shard when starting a sharded<compaction_manager>
+            // we need the getter since updateable_value is not shard-safe (#7316)
+            auto get_cm_cfg = sharded_parameter([&] {
+                return compaction_manager::config {
+                    .compaction_sched_group = compaction_manager::scheduling_group{dbcfg.compaction_scheduling_group},
+                    .maintenance_sched_group = compaction_manager::scheduling_group{dbcfg.streaming_scheduling_group},
+                    .available_memory = dbcfg.available_memory,
+                    .static_shares = cfg->compaction_static_shares,
+                    .throughput_mb_per_sec = cfg->compaction_throughput_mb_per_sec,
+                };
+            });
+            cm.start(std::move(get_cm_cfg), std::ref(abort_sources), std::ref(task_manager)).get();
+            auto stop_cm = deferred_stop(cm);
+
+            sstm.start(std::ref(*cfg)).get();
+            auto stop_sstm = deferred_stop(sstm);
+
+            db.start(std::ref(*cfg), dbcfg, std::ref(mm_notif), std::ref(feature_service), std::ref(token_metadata), std::ref(cm), std::ref(sstm), std::ref(sst_dir_semaphore), utils::cross_shard_barrier()).get();
+            auto stop_db = defer([&db] {
+                db.stop().get();
+            });
+
+            db.invoke_on_all(&replica::database::start).get();
```

Comment:
a note: Presumably 88e5b7c547bb0cd2fa33cc2dc31627b8b8953631 can be reverted (but I don't propose to check it in this PR)

Reply:
The commit does not eliminate the look, just moves the initialization early, no? Also if db indirectly depends on qctx how can it be moved before teh query_processor is created?

Reply:
db depending on query_processor sounds wrong. db should be strictly lower level IMO.

Reply:
We still need to check whether this is the case. Wrong != doesn't exist.

Reply:
I agree that we should eventually deal with it (and remove db->qp dependency if possible) but it's outside this PR's scope. The PR doesn't affect it other than making further refactors potentially easier.

Reply:
Right, if this is preexisting, then indeed it can wait.

Reply:
btw. @xemul how does `compaction_manager` use `qctx`? I cannot find the dependency.

Reply:
Ah, found it.
There's `plug_system_keyspace` called from `system_keyspace` initialization.

Then there's `unplug_system_keyspace` called when `system_keyspace` is deinitialized.

I think the dependency is fine in this case - assuming that we `plug` after `qp` is initialized and unplug before `qp` is deinitialized. (And unplugging must make sure that the dependency is no longer used when it returns; or the code who calls `unplug` must make sure of this.)

Reply:
Yes, this c.m -> qctx dependency is "gone" with plug/unplug, thus I noticed that that patch may be no longer needed

Reply:
Same for l.d.h.